rlkit/core/rl_algorithm.py

```diff
@@ -482,6 +495,128 @@ def get_extra_data_to_save(self, epoch):
             data_to_save['algorithm'] = self
         return data_to_save
 
+
+    ##### Eval stuff #####
+    def obtain_eval_paths(self, deterministic=False):
+        '''
+        collect paths with current policy
+        each transition will update the running latent context
+        the context used to condition the policy can be resampled at different intervals
+        (to enable trajectory-level or transition-level adaptation, for example)
+        '''
+        self.reset_posterior()
+        test_paths = self.eval_sampler.obtain_samples(deterministic=deterministic, resample=self.resample_z)
+        if self.sparse_rewards:
+            for p in test_paths:
+                p['rewards'] = self.env.sparsify_rewards(p['rewards'])
+        return test_paths
+
+    def collect_paths(self, idx, epoch, run):
+        self.task_idx = idx
+        self.env.reset_task(idx)
+        paths = self.obtain_eval_paths(deterministic=False)
+        goal = self.env._goal
+        for path in paths:
+            path['goal'] = goal # goal
+
+        # save the paths for visualization, only useful for point mass
+        if self.dump_eval_paths:
+            logger.save_extra_data(paths, path='eval_trajectories/task{}-epoch{}-run{}'.format(idx, epoch, run))
+
+        return paths
+
+    def _do_eval(self, indices, epoch):
+        final_returns = []
+        online_returns = []
+        for idx in indices:
+            runs, all_rets = [], []
+            for r in range(self.num_evals):
+                paths = self.collect_paths(idx, epoch, r)
+                all_rets.append([eval_util.get_average_returns([p]) for p in paths])
+                runs.append(paths)
+            all_rets = np.mean(np.stack(all_rets), axis=0) # avg return per nth rollout
+            final_returns.append(all_rets[-1])
+            online_returns.append(all_rets)
+        return final_returns, online_returns
+
+    def evaluate(self, epoch):
+        statistics = OrderedDict()
```

Comment:
do we need to do these shenanigans with the `eval_statistics` dict? It seems like every time we eval, we want to add to `self.eval_statistics` so can we just do that directly?

Reply:
while I'm thinking about this..I think in `sac.py` when we save training stats, we save them for the first batch in the epoch...which creates an offset between train and test stats. So we should probably save the last train batch stats instead? Or do an eval before we start training. And perhaps ideally average over the last few training batches in the epoch.

Reply:
I don't think the offset between stats is very significant? It's just an offset by one epoch, and I don't think we'll be looking at the statistics logged int he train loop that closely anyways, beyond a cursory glance to see that optimization is wokring.

Reply:
ok let's leave this the way it is for now